### ISSUE TITLE
rangefeed: basic measurements of performance impact of kv.changefeed.enabled

### ISSUE THREAD

**Issue by danhhz** [cockroachdb/cockroach#34836]
Turning on `kv.changefeed.enabled` adds logical operations to every raft proposal, which has a performance cost whether or not a RangeFeed is ever set up. We could do this more dynamically with some work, but that's certainly not going to make the next release. For now, just gather some preliminary info on the performance impact.

Run each of the following with and without kv.changefeed.enabled
tpcc/nodes=3/w=max (representative)
kv0/encrypt=false/nodes={1,3} (~ worst case)

**Comment by danhhz** [comment]
tpcc/nodes=3/w=max with kv.changefeed.enabled = true
```
_elapsed_______tpmC____efc__avg(ms)__p50(ms)__p90(ms)__p95(ms)__p99(ms)_pMax(ms)
 1800.0s    17306.2  96.1%    795.3    637.5   1409.3   1879.0   3758.1  47244.6
```

tpcc/nodes=3/w=max with kv.changefeed.enabled = false
```
_elapsed_______tpmC____efc__avg(ms)__p50(ms)__p90(ms)__p95(ms)__p99(ms)_pMax(ms)
 1800.0s    17572.1  97.6%    374.0    302.0    738.2    939.5   1476.4   9663.7
```

Haven't run kv tests yet, but this is pretty awful. About double across the board. cc @nvanbenschoten @tbg

**Comment by nvanbenschoten** [comment]
Yeah, that's not great. It is interesting though that we were able to still pass TPC-C at the same warehouse count.

Does anything jump out to you in any of the profiles when running with `kv.changefeed.enabled = true`? 

How does the `Disk Write Bytes` graph compare between the two runs?

**Comment by danhhz** [comment]
The clusters are long gone. If I run the tests again today, would you have a minute to poke at them?

**Comment by danhhz** [comment]
Rerunning these now. kv0 is up and running.

`kv0/encrypt=false/nodes=1` seems like any difference is in the noise. I doubled checked that the setting is correct in the two clusters. Disk Write Bytes is indeed slightly elevated (spikes between 35MB and 60MB vs spikes between 30MB and 55MB). SQL p99 is the same (both 17.8ms) ditto kv txn p99 (16.3ms) and kv txn p90 (7.6ms).

I accidentally also ran `kv0/encrypt=false/nodes=1/cpu=32` because of how the test regex works. Ditto the above: Disk Write Bytes (both are climbing but 80MB vs 70MB at a point the same duration into the test), SQL p99 (both 9.4 but rangefeed _on_ seems slightly faster dunno), kv p99 (8.4 ditto rangefeed on slightly faster), kv p90 (3.9 vs 4.1 rangefeed on faster). Rangefeed on being faster makes no sense to me. Any theories or is this just noise?

Still waiting on tpcc to restore, though I realize in retrospect I could have also tries with a smaller number of warehouses.

**Comment by danhhz** [comment]
Compared every single timeseries graph between the two `kv0/encrypt=false/nodes=1` clusters. CPU is slightly elevated with rangefeed on (up 1-2%) and Success on the Raft Log Queue graph is up (30 vs 23). The various latencies seem like the might also be _down_ just a tiny bit with rangefeed on, mirroring the cpu=32 test. I remain baffled.

**Comment by danhhz** [comment]
Okay, realized that was kv0 nodes=1. I need to try kv0 nodes=3 as well.

**Comment by tbg** [comment]
The tpcc results look weird. Same target, but only the latencies doubled?

If you leave your cluster up I can try taking a look tomorrow without having to wait for the restore.

**Comment by danhhz** [comment]
My tpcc load finished but things were all over the place during CREATE STATS, spiking above the 10s maximum resolution of our timeseries. I tried a short 10m run without them and got latencies that were more closer to what I was originally expecting for rangefeed overhead. I'm going to try a longer run with a ramp.

rangefeed=on
```
_elapsed_______tpmC____efc__avg(ms)__p50(ms)__p90(ms)__p95(ms)__p99(ms)_pMax(ms)
  600.0s    12692.3  98.7%    327.3     65.0    167.8    302.0   7784.6  16643.0
```

rangefeed=off
```
_elapsed_______tpmC____efc__avg(ms)__p50(ms)__p90(ms)__p95(ms)__p99(ms)_pMax(ms)
  600.0s    12715.4  98.9%    304.5     60.8    159.4    234.9   7784.6  13421.8
```

**Comment by danhhz** [comment]
This seems to hold up over a longer run (30m). Percentage increases work out to p50=6.7% p90=5.9% p95=7.2% p99=-11.1% avg=6.4%. We'd obviously like to get those down but 5-10% was the range I was originally expecting.

rangefeed=on
```
_elapsed_______tpmC____efc__avg(ms)__p50(ms)__p90(ms)__p95(ms)__p99(ms)_pMax(ms)
 1800.0s    12759.8  99.2%     83.3     67.1    151.0    184.5    251.7    805.3
```

rangefeed=off
```
_elapsed_______tpmC____efc__avg(ms)__p50(ms)__p90(ms)__p95(ms)__p99(ms)_pMax(ms)
 1800.0s    12741.6  99.1%     78.3     62.9    142.6    176.2    234.9    906.0
```

**Comment by nvanbenschoten** [comment]
I ran a series of `kv0` and `kv95` tests on 1 and 3 node clusters with rangefeeds off and on. Here are the results (old=`kv.rangefeed.enabled=false`, new=`kv.rangefeed.enabled=true`):
```
name                              old ops/sec  new ops/sec  delta
kv0/cores=4/nodes=1/splits=128     6.70k ± 2%   6.65k ± 2%    ~     (p=0.211 n=9+10)
kv95/cores=4/nodes=1/splits=128    15.0k ± 1%   14.9k ± 2%    ~     (p=0.393 n=10+10)
kv0/cores=4/nodes=3/splits=128     5.54k ± 1%   5.49k ± 1%  -0.88%  (p=0.001 n=10+9)
kv95/cores=4/nodes=3/splits=128    13.4k ± 0%   13.4k ± 0%    ~     (p=0.286 n=5+4)
kv0/cores=16/nodes=1/splits=128    26.2k ± 0%   26.1k ± 0%  -0.65%  (p=0.008 n=5+5)
kv95/cores=16/nodes=1/splits=128   52.8k ± 1%   52.6k ± 1%    ~     (p=0.222 n=5+5)
kv0/cores=16/nodes=3/splits=128    21.6k ± 1%   20.9k ± 5%  -3.25%  (p=0.032 n=5+5)
kv95/cores=16/nodes=3/splits=128   50.6k ± 1%   50.5k ± 1%    ~     (p=0.690 n=5+5)

name                              old p50(ms)  new p50(ms)  delta
kv0/cores=4/nodes=1/splits=128      1.20 ± 0%    1.20 ± 0%    ~     (all equal)
kv95/cores=4/nodes=1/splits=128     0.50 ± 0%    0.50 ± 0%    ~     (all equal)
kv0/cores=4/nodes=3/splits=128      1.40 ± 0%    1.40 ± 0%    ~     (all equal)
kv95/cores=4/nodes=3/splits=128     0.70 ± 0%    0.70 ± 0%    ~     (all equal)
kv0/cores=16/nodes=1/splits=128     1.20 ± 0%    1.20 ± 0%    ~     (all equal)
kv95/cores=16/nodes=1/splits=128    0.50 ± 0%    0.50 ± 0%    ~     (all equal)
kv0/cores=16/nodes=3/splits=128     1.40 ± 0%    1.40 ± 0%    ~     (all equal)
kv95/cores=16/nodes=3/splits=128    0.70 ± 0%    0.70 ± 0%    ~     (all equal)

name                              old p99(ms)  new p99(ms)  delta
kv0/cores=4/nodes=1/splits=128      2.70 ±11%    2.70 ± 4%    ~     (p=1.000 n=10+10)
kv95/cores=4/nodes=1/splits=128     1.20 ± 0%    1.20 ± 0%    ~     (all equal)
kv0/cores=4/nodes=3/splits=128      3.30 ± 0%    3.33 ± 2%    ~     (p=0.206 n=9+9)
kv95/cores=4/nodes=3/splits=128     1.50 ± 0%    1.50 ± 0%    ~     (all equal)
kv0/cores=16/nodes=1/splits=128     3.40 ± 0%    3.40 ± 0%    ~     (all equal)
kv95/cores=16/nodes=1/splits=128    4.10 ± 0%    4.10 ± 0%    ~     (all equal)
kv0/cores=16/nodes=3/splits=128     4.02 ± 3%    4.18 ±12%    ~     (p=1.000 n=5+5)
kv95/cores=16/nodes=3/splits=128    1.90 ± 0%    1.90 ± 5%    ~     (p=1.000 n=4+5)
```

All in all, things look good. The biggest performance hit is exactly where we'd expect: `kv0` on a cluster of 3, 16-core machines. The throughput drop here is **3.25%**. We see no statistically significant drop in latency due to rangefeeds being enabled across these tests.

Given that this is within our goal of enabling rangefeeds causing less than a **5%** drop to foreground traffic, I think we can consider this aspect of the feature passing release criteria (mod similar verification in #28640). I'll let @danhhz have the final say on that though.

**Comment by danhhz** [comment]
Woohoo!

I'm still curious why tpcc's latency increases look to be ~2x kv0, given that we went in thinking kv0 would be the worst case. But a 5% increase in latencies is hopefully already balanced out by rangefeeds being cheaper than polling for the changefeed itself (which is the only time we'd expect someone to have this setting on). It's certainly worthwhile to do a followup of the same three tests with a changefeed running, but I'll save that for another issue. Thanks, Nathan!
